internal/pkg/convert/node.go

```diff
@@ -0,0 +1,219 @@
+package convert
+
+import (
+	"path/filepath"
+	"sort"
+
+	"github.com/moby/buildkit/client/llb"
+	"github.com/opencontainers/go-digest"
+	"github.com/talos-systems/bldr/internal/pkg/constants"
+	"github.com/talos-systems/bldr/internal/pkg/solver"
+	"github.com/talos-systems/bldr/internal/pkg/types/v1alpha2"
+)
+
+const (
+	tmpDir = "/tmp/build"
+	pkgDir = "/pkg"
+)
+
+var defaultCopyOptions = &llb.CopyInfo{
+	CopyDirContentsOnly: true,
+	CreateDestPath:      true,
+	FollowSymlinks:      true,
+}
+
+// NodeLLB wraps PackageNode to provide LLB conversion
+type NodeLLB struct {
+	*solver.PackageNode
+
+	Graph  *GraphLLB
+	Prefix string
+}
+
+// NewNodeLLB wraps PackageNode for LLB conversion
+func NewNodeLLB(node *solver.PackageNode, graph *GraphLLB) *NodeLLB {
+	return &NodeLLB{
+		PackageNode: node,
+
+		Graph:  graph,
+		Prefix: node.Name + ":",
+	}
+}
+
+func (node *NodeLLB) base() llb.State {
+	return node.Graph.BaseImages[node.Pkg.Variant]
+}
+
+func (node *NodeLLB) install(root llb.State) llb.State {
+	if len(node.Pkg.Install) > 0 {
+		root = root.Run(
+			llb.Args(
+				append([]string{"/sbin/apk", "add", "--no-cache"}, node.Pkg.Install...)),
+			llb.WithCustomName(node.Prefix+"apk-install"),
+		).Root()
+	}
+
+	return root
+}
+
+func (node *NodeLLB) context(root llb.State) (llb.State, error) {
+	relPath := node.Pkg.BaseDir
+
+	return root.File(
+		llb.Copy(node.Graph.LocalContext, filepath.Join("/", relPath), pkgDir, defaultCopyOptions),
+		llb.WithCustomNamef(node.Prefix+"context %s -> %s", relPath, pkgDir),
+	), nil
+}
+
+func (node *NodeLLB) dependencies(root llb.State) (llb.State, error) {
+	for _, dep := range node.Pkg.Dependencies {
+		var (
+			depState llb.State
+			srcName  string
+		)
+		if dep.IsInternal() {
+			for _, depNode := range node.DependsOn {
+				if depNode.Name == dep.Stage {
+					var err error
+					depState, err = NewNodeLLB(depNode, node.Graph).Build()
+					if err != nil {
+						return llb.Scratch(), err
+					}
+					srcName = depNode.Name
+					break
+				}
+			}
+		} else {
+			depState = llb.Image(dep.Image)
+			srcName = dep.Image
+		}
+
```

Comment:
this should never happen, as this list is based on `node.InternalDeps` and it can't be missing, in the worst case it should fail if `depState` stays not initialized